# I want to check how all the data is related to each other

In [1]:
import pandas as pd

# read the dataframes 


In [25]:
accident = pd.read_csv("cleaned_data/accident.csv")
person = pd.read_csv("cleaned_data/person.csv")
vehicle = pd.read_csv("cleaned_data/vehicle.csv")

print(accident.shape)
print(vehicle.shape)
print(person.shape)

(44740, 38)
(80224, 46)
(114486, 22)


In [3]:
# let's pick one case number

# lets pick a particular case number 110215644
print("Accident details")
print(accident[(accident.CASENUM == 110215644)])


Accident details
     CASENUM  HOUR_I  MINUTE_I  RELJCT_I  ALIGN_I  PROFIL_I  SURCON_I  \
0  110215644      22        23         0        1         2         1   

   TRFCON_I  LGTCON_I  WEATHR_I  ...  NON_INVL  NUM_INJ  PED_ACC  PJ  PSU  \
0         0         2         1  ...         0        0        0   2    2   

   SPD_LIM  VEH_INVL  YEAR  NO_INJ_I  ALCHL_I  
0       99         1  2001         0        2  

[1 rows x 38 columns]


In [4]:
print("____________________")
print("Person Details for this case number")
print(person[(person.CASENUM == 110215644)])


____________________
Person Details for this case number
     CASENUM  PERNO  VEHNO  AGE_H  SEX_H  INJSEV_H  PERALC_H  WEIGHT  \
0  110215644      1      1     46      2         0         1  271.46   
1  110215644      2      1     70      2         0         0  271.46   

   PER_TYPE  SEAT_POS  ...  HOSPITAL  PER_ALCH  LOCATN  REST_SYS  IMPAIRMT  \
0         1        11  ...         0         1       0         1         0   
1         2        13  ...         0         0       0         1         0   

   ACTION  SAF_EQMT  AIRBAG  REGION  PJ  
0       0         0       9       1   2  
1       0         0       9       1   2  

[2 rows x 22 columns]


In [5]:
print("____________________")
print("Vehicle Details for this case number")
print(vehicle[(vehicle.CASENUM == 110215644)])

____________________
Vehicle Details for this case number
     CASENUM  WEIGHT  HAZM_NO  MODEL_YR  DAM_AREA  DR_ZIP_C  C_ID_NO  REGION  \
0  110215644  271.46        0      2000     13000     12456        0       1   

   HIT_RUN  MAKE  ...  OCC_INVL  PJ  PSU  P_CRASH2  SPEED  VEHNO  IMPACT_H  \
0        0    49  ...         2   2    2        87    999      1         1   

   MDLYR_I  MXVSEV_I  V_ALCH_I  
0     2000         0         2  

[1 rows x 46 columns]


For this case case number = 110215644, there was only a single vehicle involved 2 individuals 

# Let's try to analyze for this case number 160325515

In [6]:
# let's pick one case number

# accident will have a single row for sure
# lets pick a particular case number 160325515
print("Accident details")
print(accident[(accident.CASENUM == 160325515)])


Accident details
         CASENUM  HOUR_I  MINUTE_I  RELJCT_I  ALIGN_I  PROFIL_I  SURCON_I  \
34941  160325515      16        50         0        1         2         3   

       TRFCON_I  LGTCON_I  WEATHR_I  ...  NON_INVL  NUM_INJ  PED_ACC  PJ  PSU  \
34941         0         1         4  ...         0        8        0   1   11   

       SPD_LIM  VEH_INVL  YEAR  NO_INJ_I  ALCHL_I  
34941       70        23  2001         8        2  

[1 rows x 38 columns]


In [7]:
print("____________________")
print("Person Details for this case number")
print(person[(person.CASENUM == 160325515)])
print(person[(person.CASENUM == 160325515)].shape)

____________________
Person Details for this case number
         CASENUM  PERNO  VEHNO  AGE_H  SEX_H  INJSEV_H  PERALC_H  WEIGHT  \
89982  160325515      1      1     58      1         0         1  19.051   
89983  160325515      1      2     49      1         0         1  19.051   
89984  160325515      1      3     23      2         0         1  19.051   
89985  160325515      1      4     38      1         0         1  19.051   
89986  160325515      1      5     23      2         0         1  19.051   
89987  160325515      1      6     46      1         0         1  19.051   
89988  160325515      1      7     22      2         2         1  19.051   
89989  160325515      1      8     48      2         1         1  19.051   
89990  160325515      1      9     35      1         0         1  19.051   
89991  160325515      1     10     55      1         0         1  19.051   
89992  160325515      1     11     57      1         0         1  19.051   
89993  160325515      1     12 

In [8]:
print("____________________")
print("Vehicle Details for this case number")
print(vehicle[(vehicle.CASENUM == 160325515)])

____________________
Vehicle Details for this case number
         CASENUM  WEIGHT  HAZM_NO  MODEL_YR  DAM_AREA  DR_ZIP_C  C_ID_NO  \
62598  160325515  19.051        0      1997     10000     49234        0   
62599  160325515  19.051        0      1994     10000     53209        0   
62600  160325515  19.051        0      1991     10000     49002        0   
62601  160325515  19.051        0      2001     40000     78840   658570   
62602  160325515  19.051        0      2000     40000     49201        0   
62603  160325515  19.051        0      1999     40000     60433   146880   
62604  160325515  19.051        0      1999     10000     48170        0   
62605  160325515  19.051        0      1997     10000     49001        0   
62606  160325515  19.051        0      1992     20000     48504        0   
62607  160325515  19.051        0      1995     10000     63351   438719   
62608  160325515  19.051        0      1997     34000     49201        0   
62609  160325515  19.051     9

## Question:- Can there be multiple case number in accident data?

In [9]:
# for accident we have a single CASENUMBER for particlar 
(accident.CASENUM.value_counts() > 1).any()


False

# Answer - No

In [10]:
# Let's check the same for vehicle 
# I am assuming that for a case number there can be multiple vehicle involved

(vehicle.CASENUM.value_counts() > 1).any()

True

In [11]:
duplicate_cases = vehicle.CASENUM.value_counts()
duplicate_cases = duplicate_cases[duplicate_cases > 1]  # Filter cases with count > 1


In [12]:
duplicate_cases

CASENUM
160325515    23
130236944    10
180809192     9
162917987     9
132460321     8
             ..
122275543     2
130245938     2
122275465     2
130235824     2
130235513     2
Name: count, Length: 30568, dtype: int64

In [13]:

print("____________________")
print("Person Details for this case number")
print(person[(person.CASENUM == 130236944)])
print(person[(person.CASENUM == 130236944)].shape)

____________________
Person Details for this case number
         CASENUM  PERNO  VEHNO  AGE_H  SEX_H  INJSEV_H  PERALC_H  WEIGHT  \
30808  130236944      1      1     42      1         0         1   2.701   
30809  130236944      1      2     23      1         4         1   2.701   
30810  130236944      1      3     19      2         1         1   2.701   
30811  130236944      2      3     17      2         1         0   2.701   
30812  130236944      1      4     48      1         2         1   2.701   
30813  130236944      1      5     20      2         1         1   2.701   
30814  130236944      1      6     50      1         0         1   2.701   
30815  130236944      1      7     20      1         1         1   2.701   
30816  130236944      2      7     41      1         1         0   2.701   
30817  130236944      1      8     52      2         2         1   2.701   
30818  130236944      1      9     52      2         2         1   2.701   
30819  130236944      2      9 

In [14]:
print("____________________")
print("Vehicle Details for this case number")
print(vehicle[(vehicle.CASENUM == 130236944)])

____________________
Vehicle Details for this case number
         CASENUM  WEIGHT  HAZM_NO  MODEL_YR  DAM_AREA  DR_ZIP_C  C_ID_NO  \
21367  130236944   2.701        0      1988     19999     48186   994131   
21368  130236944   2.701        0      1993     39999     48125        0   
21369  130236944   2.701        0      1983     19999     48180        0   
21370  130236944   2.701        0      2001     39999     48146        0   
21371  130236944   2.701        0      1990     12999     48125        0   
21372  130236944   2.701        0      1989     19999     48180        0   
21373  130236944   2.701        0      1990     13999     48179        0   
21374  130236944   2.701        0      1994     19999     48188        0   
21375  130236944   2.701        0      2000     13999     48126        0   
21376  130236944   2.701        0      1991     10000     48101        0   

       REGION  HIT_RUN  MAKE  ...  OCC_INVL  PJ  PSU  P_CRASH2  SPEED  VEHNO  \
21367       2        0   

 # Ok let's try to merge the data now

In [15]:
# we'll merge accident and vehicle data

common_columns = list(set(accident.columns) & set(vehicle.columns))

# Print the common columns
print("Common columns:", common_columns)


Common columns: ['WEIGHT', 'PJ', 'CASENUM', 'PSU', 'REGION']


In [63]:
# we'll merge accident and vehicle data

common_columns = list(set(accident.columns) & set(person.columns))

# Print the common columns
print("Common columns:", common_columns)


Common columns: ['CASENUM', 'WEIGHT', 'PJ', 'REGION']


In [64]:
# we'll merge accident and vehicle data

common_columns = list(set(vehicle.columns) & set(person.columns))

# Print the common columns
print("Common columns:", common_columns)


Common columns: ['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']


Region column analysis

In [16]:
"""print(vehicle[(vehicle.CASENUM == 160325515)]['PSU'].unique())
print(accident[(accident.CASENUM == 160325515)]['PSU'].unique())


print(vehicle[(vehicle.CASENUM == 160325515)]['PJ'].unique())
print(accident[(accident.CASENUM == 160325515)]['PJ'].unique())

print(vehicle[(vehicle.CASENUM == 160325515)]['REGION'].unique())
print(accident[(accident.CASENUM == 160325515)]['REGION'].unique())
"""
accident_sample = accident[["CASENUM","REGION"]]
vehicle_sample = vehicle[["CASENUM","REGION"]]

# Group vehicle_sample by 'CASENUM' and stack the 'REGION' values into a list
grouped_vehicle_sample = vehicle_sample.groupby('CASENUM')['REGION'].agg(lambda x: set(x)).reset_index()

# Display the result
#print(grouped_vehicle_sample)

# Filter rows where the length of the 'REGION' set is greater than 1
multiple_regions = grouped_vehicle_sample[grouped_vehicle_sample['REGION'].apply(len) > 1]

# Display the result
print(multiple_regions)



Empty DataFrame
Columns: [CASENUM, REGION]
Index: []


PJ column analysis

In [17]:

accident_sample = accident[["CASENUM","PJ"]]
vehicle_sample = vehicle[["CASENUM","PJ"]]

# Group vehicle_sample by 'CASENUM' and stack the 'REGION' values into a list
grouped_vehicle_sample = vehicle_sample.groupby('CASENUM')['PJ'].agg(lambda x: set(x)).reset_index()

# Display the result
#print(grouped_vehicle_sample)

# Filter rows where the length of the 'REGION' set is greater than 1
multiple_regions = grouped_vehicle_sample[grouped_vehicle_sample['PJ'].apply(len) > 1]

# Display the result
print(multiple_regions)



Empty DataFrame
Columns: [CASENUM, PJ]
Index: []


PSU column analysis

In [18]:

accident_sample = accident[["CASENUM","PSU"]]
vehicle_sample = vehicle[["CASENUM","PSU"]]

# Group vehicle_sample by 'CASENUM' and stack the 'REGION' values into a list
grouped_vehicle_sample = vehicle_sample.groupby('CASENUM')['PSU'].agg(lambda x: set(x)).reset_index()

# Display the result
#print(grouped_vehicle_sample)

# Filter rows where the length of the 'REGION' set is greater than 1
multiple_regions = grouped_vehicle_sample[grouped_vehicle_sample['PSU'].apply(len) > 1]

# Display the result
print(multiple_regions)



Empty DataFrame
Columns: [CASENUM, PSU]
Index: []


so in vehicle and accident data ['PSU', 'REGION', 'PJ', 'WEIGHT', 'CASENUM'] columns resembles the same values

In [19]:
# taking accident on left and vehicle on right we can do a left join on ['PSU', 'REGION', 'PJ', 'WEIGHT', 'CASENUM'] these columns 

# Perform a left join on the specified columns
merged_data_accident_vehicle = pd.merge(accident, vehicle, on=['PSU', 'REGION', 'PJ', 'WEIGHT', 'CASENUM'], how='left')

print(merged_data_accident_vehicle.shape)
# Display the result


(80224, 79)


In [20]:
merged_data_accident_vehicle.to_csv("cleaned_data/merged_data_accident_vehicle.csv",index=False)

In [ ]:
# check for null values
t1 = tuple(zip(list(merged_data_accident_vehicle.columns),list(merged_data_accident_vehicle.isna().sum())))
print([(col, missing) for col, missing in t1 if missing > 0])


# There are no null values

[]


## we have 80224 rows which is equal to vehicle shape, thus we can say that the data was joined completely without anything left in the vehicles dataframe

now we need to merge the `merged_data_accident_vehicle` with `person` table

In [21]:
# we'll merge accident and vehicle data

common_columns = list(set(merged_data_accident_vehicle.columns) & set(person.columns))

# Print the common columns
print("Common columns:", common_columns)


Common columns: ['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']


In [22]:

accident_sample = merged_data_accident_vehicle[["CASENUM","PJ"]]
vehicle_sample = person[["CASENUM","PJ"]]

# Group vehicle_sample by 'CASENUM' and stack the 'REGION' values into a list
grouped_vehicle_sample = vehicle_sample.groupby('CASENUM')['PJ'].agg(lambda x: set(x)).reset_index()

# Display the result
#print(grouped_vehicle_sample)

# Filter rows where the length of the 'REGION' set is greater than 1
multiple_regions = grouped_vehicle_sample[grouped_vehicle_sample['PJ'].apply(len) > 1]

# Display the result
print(multiple_regions)



Empty DataFrame
Columns: [CASENUM, PJ]
Index: []


In [23]:

accident_sample = merged_data_accident_vehicle[["CASENUM","REGION"]]
vehicle_sample = person[["CASENUM","REGION"]]

# Group vehicle_sample by 'CASENUM' and stack the 'REGION' values into a list
grouped_vehicle_sample = vehicle_sample.groupby('CASENUM')['REGION'].agg(lambda x: set(x)).reset_index()

# Display the result
#print(grouped_vehicle_sample)

# Filter rows where the length of the 'REGION' set is greater than 1
multiple_regions = grouped_vehicle_sample[grouped_vehicle_sample['REGION'].apply(len) > 1]

# Display the result
print(multiple_regions)



Empty DataFrame
Columns: [CASENUM, REGION]
Index: []


ok I can see both region and pj are have same data values, thus we can use all these to join the data

In [54]:
# taking accident on left and vehicle on right we can do a left join on ['PSU', 'REGION', 'PJ', 'WEIGHT', 'CASENUM'] these columns 

# Perform a left join on the specified columns
final_merge = pd.merge(merged_data_accident_vehicle, person, on=['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION'], how='left')

print(final_merge.shape)
# Display the result


(112154, 96)


In [ ]:
114486-112154

2332

In [33]:
# check for null values
t1 = tuple(zip(list(final_merge.columns),list(final_merge.isna().sum())))
print([(col, missing) for col, missing in t1 if missing > 0])


# There are no null values

[('PERNO', 159), ('AGE_H', 159), ('SEX_H', 159), ('INJSEV_H', 159), ('PERALC_H', 159), ('PER_TYPE', 159), ('SEAT_POS', 159), ('EJECT', 159), ('SEX', 159), ('HOSPITAL', 159), ('PER_ALCH', 159), ('LOCATN', 159), ('REST_SYS', 159), ('IMPAIRMT', 159), ('ACTION', 159), ('SAF_EQMT', 159), ('AIRBAG', 159)]


In [61]:
# Filter rows where the 'PERSONID' is NaN (indicating no match)
unjoined_rows = final_merge[final_merge['PERNO'].isna()]

# Display the unjoined rows
print(unjoined_rows)

          CASENUM  HOUR_I  MINUTE_I  RELJCT_I  ALIGN_I  PROFIL_I  SURCON_I  \
1496    110218576      20        35         2        1         1         1   
1638    110218850      18        35         1        1         2         1   
1890    110219251      17        19         0        1         2         1   
2831    110672913       6        23         0        1         1         1   
3486    110676327      16         0         0        1         1         1   
...           ...     ...       ...       ...      ...       ...       ...   
106680  180907055      21         0         3        1         2         2   
107453  181111786       5        41         0        1         1         2   
108082  182128549      10        20         0        1         1         1   
111476  188466213      22         3         0        1         1         1   
111748  198143005       6        30         2        1         1         2   

        TRFCON_I  LGTCON_I  WEATHR_I  ...  EJECT  SEX  HOSPITAL

In [58]:
# get case numbers of unjoined rows

print("For these case numbers we actually dont have any information in person column")
unjoined_rows["CASENUM"].unique()

For these case numbers we actually dont have any information in person column


array([110218576, 110218850, 110219251, 110672913, 110676327, 110756562,
       110757472, 113421563, 114920633, 114924743, 115319148, 116657805,
       117032669, 117033448, 117033827, 117576518, 117577465, 117579238,
       117579326, 117579337, 117581005, 117581941, 117583844, 117588748,
       117589168, 118832502, 120410830, 120412365, 120709897, 120851067,
       120852123, 120852953, 120854449, 120857183, 120866983, 120872395,
       121229052, 121230048, 121230435, 121970681, 121972063, 122274519,
       122280189, 122285702, 129926203, 130020558, 130021703, 130226355,
       130230035, 130230347, 130237806, 130238759, 130240032, 130244454,
       130246639, 130464392, 130599448, 130605297, 130610116, 130610980,
       130614259, 130631564, 130632071, 130734063, 131114885, 131246081,
       131251250, 131356980, 131357864, 131361884, 131407843, 131408948,
       131409110, 131572947, 131582931, 131618271, 131621294, 131720780,
       131722196, 131722230, 131724894, 131725683, 

In [62]:
print("person")
print(person[(person.CASENUM == 110218576)][['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']])


person
      WEIGHT  PJ    CASENUM  VEHNO  REGION
1509  89.853  10  110218576      2       1


In [59]:
caseno = 110218576
print(caseno)
print("vehicle")
print(vehicle[(vehicle.CASENUM == caseno)][['WEIGHT', 'CASENUM', 'VEHNO', 'REGION','PJ']])
print("--------------------------------")
print("accident")
print(accident[(accident.CASENUM == caseno)][['WEIGHT', 'PJ', 'CASENUM', 'REGION']])
print("--------------------------------")
print("person")
print(person[(person.CASENUM == caseno)][['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']])
print("--------------------------------")
print("merged_data_accident_vehicle")
print(merged_data_accident_vehicle[(merged_data_accident_vehicle.CASENUM == caseno)][['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']])
print("--------------------------------")
print("final_merge")
print(final_merge[(final_merge.CASENUM == caseno)][['WEIGHT', 'PJ', 'CASENUM', 'VEHNO', 'REGION']])


110218576
vehicle
      WEIGHT    CASENUM  VEHNO  REGION  PJ
1047  89.853  110218576      1       1  10
1048  89.853  110218576      2       1  10
--------------------------------
accident
     WEIGHT  PJ    CASENUM  REGION
677  89.853  10  110218576       1
--------------------------------
person
      WEIGHT  PJ    CASENUM  VEHNO  REGION
1509  89.853  10  110218576      2       1
--------------------------------
merged_data_accident_vehicle
      WEIGHT  PJ    CASENUM  VEHNO  REGION
1047  89.853  10  110218576      1       1
1048  89.853  10  110218576      2       1
--------------------------------
final_merge
      WEIGHT  PJ    CASENUM  VEHNO  REGION
1496  89.853  10  110218576      1       1
1497  89.853  10  110218576      2       1


In [68]:
# let's just remove the null values

final_merge_ = final_merge.dropna()

In [69]:
final_merge_.columns

Index(['CASENUM', 'HOUR_I', 'MINUTE_I', 'RELJCT_I', 'ALIGN_I', 'PROFIL_I',
       'SURCON_I', 'TRFCON_I', 'LGTCON_I', 'WEATHR_I', 'MANCOL_I', 'EVENT1_I',
       'WEIGHT', 'MONTH', 'WEEKDAY', 'REGION', 'LAND_USE', 'INT_HWY',
       'REL_RWY', 'TRAF_WAY', 'ALIGN', 'PROFILE', 'WEATHER', 'SCHL_BUS',
       'WRK_ZONE', 'ALCOHOL', 'NUM_LAN', 'HOUR', 'NON_INVL', 'NUM_INJ',
       'PED_ACC', 'PJ', 'PSU', 'SPD_LIM', 'VEH_INVL', 'YEAR', 'NO_INJ_I',
       'ALCHL_I', 'HAZM_NO', 'MODEL_YR', 'DAM_AREA', 'DR_ZIP_C', 'C_ID_NO',
       'HIT_RUN', 'MAKE', 'BODY_TYP', 'SPEC_USE', 'EMCY_USE', 'FACTOR',
       'TRAILER', 'JACKNIFE', 'FIRE', 'VEH_SEV', 'TOWED', 'P_CRASH1',
       'VEH_ROLE', 'ACC_TYPE', 'IMPACT', 'P_CRASH3', 'PCRASH4', 'PCRASH5',
       'ROLLOVER', 'VIOLATN', 'VIS_OBSC', 'DR_DSTRD', 'SPEEDREL', 'V_EVENT',
       'VEH_ALCH', 'MAX_VSEV', 'MHENUM', 'NUMOCCS', 'OCC_INVL', 'P_CRASH2',
       'SPEED', 'VEHNO', 'IMPACT_H', 'MDLYR_I', 'MXVSEV_I', 'V_ALCH_I',
       'PERNO', 'AGE_H', 'SEX_H', 'INJS

In [70]:
import pandas as pd

# Dictionary with column names as keys and values to be removed
remove_values = {
    'LGTCON_I': 8, 'WEATHR_I': 9, 'MANCOL_I': 9, 'WEEKDAY': 9, 'LAND_USE': 9, 
    'MAN_COL': 9, 'INT_HWY': 9, 'REL_JCT': 98, 'REL_RWY': 98, 'ALIGN': 9, 
    'PROFILE': 9, 'SUR_COND': 9, 'TRAF_CON': 9, 'LGHT_CON': 8, 'WEATHER': 9, 
    'ALCOHOL': 9, 'NUM_LAN': 9, 'EVENT1': 99, 'HOUR': 99, 'MINUTE': 99, 
    'NUM_INJ': 99, 'VEHNO': 99, 'INJSEV_H': 99, 'SEAT_H': 8, 'PER_TYPE': 99, 
    'SEAT_POS': 3, 'EJECT': 9, 'SEX': 9, 'HOSPITAL': 9, 'PER_ALCH': 99, 
    'LOCATN': 9, 'REST_SYS': 9, 'PER_DRUG': 99, 'IMPAIRMT': 99, 'ACTION': 9, 
    'SAF_EQMT': 9, 'REGION': 99, 'STR_VEH': 3, 'VIN': 999, 'HAZM_NO': 999, 
    'DAM_AREA': 9, 'HIT_RUN': 9, 'SPEC_USE': 99, 'EMCY_USE': 9, 'FACTOR': 99, 
    'TRAILER': 6, 'VEH_SEV': 9, 'TOWED': 9, 'P_CRASH1': 99, 'VEH_ROLE': 9, 
    'P_CRASH3': 99, 'PCRASH5': 99, 'CARG_TYP': 99, 'DR_PRES': 9, 'VIOLATN': 96, 
    'VIS_OBSC': 96, 'DRMAN_AV': 96, 'DR_DSTRD': 96, 'SPEEDREL': 9, 'VEH_ALCH': 9, 
    'MAX_VSEV': 9, 'NUMOCCS': 999, 'OCC_INVL': 97, 'SPEED': 999, 'VROLE_I': 9, 
    'V_EVNT_H': 99
}

# Loop over the dictionary to filter rows based on column names and values
for column, value in remove_values.items():
    if column in final_merge_.columns:
        final_merge_cleaned = final_merge_[final_merge_[column] != value]

# The filtered dataframe is now in final_merge
final_merge_cleaned.shape

(50967, 96)

In [71]:
final_merge_cleaned.to_csv("cleaned_data/final_merge_cleaned.csv",index=False)

In [73]:
final_merge_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50967 entries, 5 to 112153
Data columns (total 96 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   CASENUM   50967 non-null  int64  
 1   HOUR_I    50967 non-null  int64  
 2   MINUTE_I  50967 non-null  int64  
 3   RELJCT_I  50967 non-null  int64  
 4   ALIGN_I   50967 non-null  int64  
 5   PROFIL_I  50967 non-null  int64  
 6   SURCON_I  50967 non-null  int64  
 7   TRFCON_I  50967 non-null  int64  
 8   LGTCON_I  50967 non-null  int64  
 9   WEATHR_I  50967 non-null  int64  
 10  MANCOL_I  50967 non-null  int64  
 11  EVENT1_I  50967 non-null  int64  
 12  WEIGHT    50967 non-null  float64
 13  MONTH     50967 non-null  int64  
 14  WEEKDAY   50967 non-null  int64  
 15  REGION    50967 non-null  int64  
 16  LAND_USE  50967 non-null  int64  
 17  INT_HWY   50967 non-null  int64  
 18  REL_RWY   50967 non-null  int64  
 19  TRAF_WAY  50967 non-null  int64  
 20  ALIGN     50967 non-null  int64 

In [85]:
%pip install FLAML

  Using cached FLAML-2.3.4-py3-none-any.whl.metadata (16 kB)
Using cached FLAML-2.3.4-py3-none-any.whl (314 kB)
Note: you may need to restart the kernel to use updated packages.


In [110]:
%pip install catboost
%pip install --upgrade datasets

Note: you may need to restart the kernel to use updated packages.
  Attempting uninstall: datasets
    Found existing installation: datasets 3.2.0
    Uninstalling datasets-3.2.0:
      Successfully uninstalled datasets-3.2.0
Note: you may need to restart the kernel to use updated packages.


In [126]:
import numpy as np
X,y = final_merge_cleaned.drop(['CASENUM','INJSEV_H'],axis=1),final_merge_cleaned[['INJSEV_H']]
X= np.array(X)
y = np.array(y)

In [127]:
from sklearn.model_selection import train_test_split
from flaml import AutoML
from sklearn.metrics import f1_score


In [128]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [129]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((40773, 94), (10194, 94), (40773, 1), (10194, 1))

In [130]:

auto_ml = AutoML()

# Set up AutoML configuration
settings = {
    "time_budget": 60,  # Time budget in seconds
    "metric": 'accuracy',  # Use F1 score as the metric
    "task": 'classification',  # Classification task
    "log_file_name": "automl.log",  # Optional log file for debugging
    "estimator_list": ["xgboost","rf","extra_tree","xgb_limitdepth","sgd","catboost","lrl1"]  # Specify models
}
#["xgboost","rf","extra_tree","xgb_limitdepth","sgd","catboost","lrl1"]

# Train AutoML model
auto_ml.fit(X_train=X_train, y_train=y_train, **settings)

# Make predictions
y_pred = auto_ml.predict(X_test)

# Evaluate the model
accuracy = f1_score(y_test, y_pred,average='weighted')



[flaml.automl.logger: 03-23 15:36:37] {1728} INFO - task = classification
[flaml.automl.logger: 03-23 15:36:37] {1739} INFO - Evaluation method: holdout
[flaml.automl.logger: 03-23 15:36:37] {1838} INFO - Minimizing error metric: 1-accuracy
[flaml.automl.logger: 03-23 15:36:37] {1955} INFO - List of ML learners in AutoML Run: ['xgboost', 'rf', 'extra_tree', 'xgb_limitdepth', 'sgd', 'catboost', 'lrl1']
[flaml.automl.logger: 03-23 15:36:37] {2258} INFO - iteration 0, current learner xgboost
[flaml.automl.logger: 03-23 15:36:37] {2393} INFO - Estimated sufficient time budget=986s. Estimated necessary time budget=15s.
[flaml.automl.logger: 03-23 15:36:37] {2442} INFO -  at 0.2s,	estimator xgboost's best error=0.0649,	best estimator xgboost's best error=0.0649
[flaml.automl.logger: 03-23 15:36:37] {2258} INFO - iteration 1, current learner sgd
[flaml.automl.logger: 03-23 15:36:38] {2442} INFO -  at 0.4s,	estimator sgd's best error=0.2871,	best estimator xgboost's best error=0.0649
[flaml.au

In [ ]:
print("F1-Score", accuracy)

0.9550158118584249


In [132]:
print(auto_ml.model)

In [ ]:
best_model = auto_ml.model
best_params = best_model.get_params() if hasattr(best_model, 'get_params') else "No parameters available"


{'n_estimators': 46, 'max_leaves': 6, 'min_child_weight': 3.4353192705502913, 'learning_rate': 0.399189341291266, 'subsample': 0.903512345912589, 'colsample_bylevel': 0.7563793731764131, 'colsample_bytree': 0.7647693832410042, 'reg_alpha': 0.03542523638424809, 'reg_lambda': 0.10434057752736216, 'n_jobs': -1, 'max_depth': 0, 'grow_policy': 'lossguide', 'tree_method': 'hist', 'verbosity': 0, 'task': <flaml.automl.task.generic_task.GenericTask object at 0x3398030b0>, '_estimator_type': 'classifier'}


In [135]:
# Print the best model parameters one by one
if best_params:
    print("Best Model Parameters:")
    for param, value in best_params.items():
        print(f"{param}: {value}")
else:
    print("No parameters available for the best model.")

Best Model Parameters:
n_estimators: 46
max_leaves: 6
min_child_weight: 3.4353192705502913
learning_rate: 0.399189341291266
subsample: 0.903512345912589
colsample_bylevel: 0.7563793731764131
colsample_bytree: 0.7647693832410042
reg_alpha: 0.03542523638424809
reg_lambda: 0.10434057752736216
n_jobs: -1
max_depth: 0
grow_policy: lossguide
tree_method: hist
verbosity: 0
task: multiclass
_estimator_type: classifier
